File: src/world.js
import {
    VIEW_W, VIEW_H, GROUND_Y, CEILING_Y,
    PIPE_W, PIPE_SPACING, PIPE_GAP_START, PIPE_GAP_MIN,
    SCROLL_BASE, SCROLL_MAX, MUSHROOM_CHANCE, MUSHROOM_COOLDOWN,
    TRIPLE_DURATION, TRIPLE_MULT, COMBO_WINDOW,
} from './config.js';
import { Bird, Pipe, Coin, Mushroom, Particle, FloatText } from './entities.js';
import { C } from './palette.js';
import { CLOUD, drawText, drawTextCentered, measureText } from './sprites.js';
import { sfx } from './audio.js';
import { FX } from './fx.js';

function lerp(a, b, t) { return a + (b - a) * t; }
function clamp(v, a, b) { return v < a ? a : (v > b ? b : v); }

export class World {
    constructor() {
        this.fx = new FX();
        this.reset();
        // Parallax layers (seeded random for mountains + clouds)
        this.clouds = [];
        for (let i = 0; i < 6; i++) {
            this.clouds.push({
                x: Math.random() * VIEW_W,
                y: 20 + Math.random() * 120,
                speed: 0.12 + Math.random() * 0.22,
            });
        }
        this.mountains = this._buildMountains(VIEW_W * 2);
        this.mountainScroll = 0;
        this.groundScroll = 0;
    }

    _buildMountains(width) {
        // Simple seeded noise silhouette.
        const heights = new Array(width);
        let a = 0.6, b = 1.7, c = 3.3;
        for (let x = 0; x < width; x++) {
            const n =
                Math.sin(x * 0.03 + a) * 0.55 +
                Math.sin(x * 0.07 + b) * 0.3 +
                Math.sin(x * 0.17 + c) * 0.15;
            heights[x] = 120 + n * 22;
        }
        return heights;
    }

    reset() {
        this.bird = new Bird();
        this.pipes = [];
        this.coins = [];
        this.mushrooms = [];
        this.particles = [];
        this.floatTexts = [];
        this.score = 0;
        this.coinCount = 0;
        this.combo = 0;
        this.comboTimer = 0;
        this.tripleTimer = 0;
        this.mushroomCooldown = 0;
        this.spawnX = VIEW_W + 40;
        this.elapsed = 0;
        this.gameOver = false;
        this.started = false;
    }

    scrollSpeed() {
        const t = clamp(this.score / 40, 0, 1);
        return lerp(SCROLL_BASE, SCROLL_MAX, t);
    }

    pipeGap() {
        const t = clamp(this.score / 35, 0, 1);
        return Math.round(lerp(PIPE_GAP_START, PIPE_GAP_MIN, t));
    }

    flap() {
        if (this.gameOver) return;
        if (!this.started) this.started = true;
        this.bird.flap();
        // Flap puff particles
        for (let i = 0; i < 3; i++) {
            this.particles.push(new Particle(
                this.bird.x - 4, this.bird.y + 4,
                -20 - Math.random() * 20, 10 + Math.random() * 20,
                0.35, C.CREAM, 1, 0
            ));
        }
        sfx.flap();
    }

    spawnPipeRun() {
        const gap = this.pipeGap();
        const minTop = 22;
        const maxTop = GROUND_Y - gap - 22;
        const gapY = Math.floor(minTop + Math.random() * (maxTop - minTop));
        const pipe = new Pipe(this.spawnX, gapY, gap);
        this.pipes.push(pipe);

        // Spawn collectibles inside/near this gap.
        this._spawnCoinPattern(pipe);

        if (this.mushroomCooldown <= 0 && Math.random() < MUSHROOM_CHANCE) {
            const my = gapY + gap / 2 + (Math.random() * 10 - 5);
            this.mushrooms.push(new Mushroom(this.spawnX + PIPE_W / 2, my));
            this.mushroomCooldown = MUSHROOM_COOLDOWN;
        }

        this.spawnX += PIPE_SPACING;
    }

    _spawnCoinPattern(pipe) {
        const patterns = ['arc', 'line', 'cluster'];
        const kind = patterns[(Math.random() * patterns.length) | 0];
        const cx = pipe.x + PIPE_W / 2;
        const cy = pipe.gapY + pipe.gapH / 2;

        if (kind === 'line') {
            const n = 4 + ((Math.random() * 3) | 0);
            for (let i = 0; i < n; i++) {
                this.coins.push(new Coin(cx - (n - 1) * 5 + i * 10, cy));
            }
        } else if (kind === 'arc') {
            const n = 5;
            for (let i = 0; i < n; i++) {
                const t = i / (n - 1);
                const ax = cx - 18 + t * 36;
                const ay = cy - Math.sin(t * Math.PI) * 18;
                this.coins.push(new Coin(ax, ay));
            }
        } else {
            // cluster: small bunch in the middle
            this.coins.push(new Coin(cx, cy));
            this.coins.push(new Coin(cx - 7, cy - 4));
            this.coins.push(new Coin(cx + 7, cy - 4));
            this.coins.push(new Coin(cx - 7, cy + 4));
            this.coins.push(new Coin(cx + 7, cy + 4));
        }
    }

    update(dt) {
        this.elapsed += dt;
        if (this.started && !this.gameOver) {
            const scroll = this.scrollSpeed();
            // Spawn
            if (this.pipes.length === 0 || this.pipes[this.pipes.length - 1].x < this.spawnX - PIPE_SPACING) {
                this.spawnPipeRun();
            }

            // Move world
            for (const p of this.pipes) p.update(dt, scroll);
            for (const c of this.coins) c.update(dt, scroll);
            for (const m of this.mushrooms) m.update(dt, scroll);
            this.groundScroll = (this.groundScroll + scroll * dt) % 8;
            this.mountainScroll = (this.mountainScroll + scroll * 0.15 * dt);
            for (const cl of this.clouds) {
                cl.x -= scroll * cl.speed * dt;
                if (cl.x < -20) { cl.x = VIEW_W + Math.random() * 40; cl.y = 20 + Math.random() * 120; }
            }
            this.spawnX -= scroll * dt;

            this.mushroomCooldown = Math.max(0, this.mushroomCooldown - dt);
            if (this.tripleTimer > 0) this.tripleTimer = Math.max(0, this.tripleTimer - dt);
            if (this.comboTimer > 0) {
                this.comboTimer -= dt;
                if (this.comboTimer <= 0) this.combo = 0;
            }
        }

        // Bird always updates (even at game-over so it falls).
        this.bird.update(dt);

        // Particles + float texts
        for (const p of this.particles) p.update(dt);
        this.particles = this.particles.filter(p => p.alive);
        for (const t of this.floatTexts) t.update(dt);
        this.floatTexts = this.floatTexts.filter(t => t.alive);

        // Ground / ceiling clamp
        if (this.bird.y + this.bird.r >= GROUND_Y) {
            this.bird.y = GROUND_Y - this.bird.r;
            if (this.started && !this.gameOver) this._kill('ground');
            this.bird.vy = 0;
        }
        if (this.bird.y - this.bird.r < CEILING_Y) {
            this.bird.y = CEILING_Y + this.bird.r;
            if (this.bird.vy < 0) this.bird.vy = 0;
        }

        if (this.started && !this.gameOver) this._collisions();

        // Cleanup
        this.pipes = this.pipes.filter(p => p.x + PIPE_W > -4);
        this.coins = this.coins.filter(c => c.alive && c.x > -10);
        this.mushrooms = this.mushrooms.filter(m => m.alive && m.x > -10);

        this.fx.update(dt);
    }

    _collisions() {
        const b = this.bird;

        // Pipes: check top + bottom of each pipe.
        for (const p of this.pipes) {
            // Score-on-pass
            if (!p.passed && p.x + PIPE_W < b.x) {
                p.passed = true;
                this.score += 1;
                sfx.score();
            }
            const t = p.topRect, bo = p.botRect;
            if (b.hitsRect(t.x, t.y, t.w, t.h) || b.hitsRect(bo.x, bo.y, bo.w, bo.h)) {
                this._kill('pipe');
                return;
            }
        }

        // Coins
        for (const c of this.coins) {
            if (!c.alive) continue;
            if (c.hits(b)) {
                c.alive = false;
                this._collectCoin(c);
            }
        }

        // Mushrooms
        for (const m of this.mushrooms) {
            if (!m.alive) continue;
            if (m.hits(b)) {
                m.alive = false;
                this._collectMushroom(m);
            }
        }
    }

    _collectCoin(c) {
        const mult = this.tripleTimer > 0 ? TRIPLE_MULT : 1;
        const gain = 1 * mult;
        this.score += gain;
        this.coinCount += 1;
        this.combo += 1;
        this.comboTimer = COMBO_WINDOW;

        // Particles
        for (let i = 0; i < 7; i++) {
            const ang = Math.random() * Math.PI * 2;
            const sp = 30 + Math.random() * 50;
            this.particles.push(new Particle(
                c.x, c.y,
                Math.cos(ang) * sp, Math.sin(ang) * sp,
                0.45, i % 2 ? C.GOLD : C.CREAM, 1, 60
            ));
        }
        this.floatTexts.push(new FloatText(c.x, c.y - 6, '+' + gain, mult > 1 ? C.ORANGE : C.GOLD));
        this.fx.flash(C.CREAM, 0.05);
        sfx.coin();
        if (this.combo >= 3) sfx.combo(this.combo);
    }

    _collectMushroom(m) {
        this.tripleTimer = TRIPLE_DURATION;
        this.fx.shake(2, 0.12);
        this.fx.slowTime(0.35, 0.22);
        this.fx.flash(C.ORANGE, 0.12);
        // Radial burst
        for (let i = 0; i < 22; i++) {
            const ang = (i / 22) * Math.PI * 2;
            const sp = 40 + Math.random() * 40;
            this.particles.push(new Particle(
                m.x, m.y,
                Math.cos(ang) * sp, Math.sin(ang) * sp,
                0.7, i % 3 === 0 ? C.WHITE : (i % 2 ? C.ORANGE : C.CREAM), 1, 20
            ));
        }
        this.floatTexts.push(new FloatText(m.x, m.y - 8, '3X POWER!', C.ORANGE));
        sfx.mushroom();
    }

    _kill(reason) {
        if (this.gameOver) return;
        this.gameOver = true;
        this.bird.alive = false;
        this.fx.shake(6, 0.35);
        this.fx.flash(C.CRIMSON, 0.18);
        for (let i = 0; i < 16; i++) {
            const ang = Math.random() * Math.PI * 2;
            const sp = 40 + Math.random() * 60;
            this.particles.push(new Particle(
                this.bird.x, this.bird.y,
                Math.cos(ang) * sp, Math.sin(ang) * sp,
                0.9, i % 2 ? C.CRIMSON : C.ORANGE, 1, 200
            ));
        }
        sfx.hit();
    }
}

File: skybit/world.py
import math
import random
import pyxel
from .config import (GROUND_Y, BIRD_X, PIPE_W, PIPE_SPACING,
                     GAP_START, GAP_MIN, SCROLL_BASE, SCROLL_MAX,
                     MUSHROOM_CHANCE, MUSHROOM_COOLDOWN, TRIPLE_DURATION,
                     COMBO_WINDOW)
from .entities import Bird, Pipe, Coin, Mushroom, Particle, FloatText
from .palette import DKRED, ORNG, GOLD, CREAM, SKY, RED, WHITE, WGRN


def _lerp(a, b, t):
    return a + (b - a) * max(0.0, min(1.0, t))


class World:
    def __init__(self):
        self.bird = Bird()
        self._reset_state()
        self._mountain_scroll = 0.0
        self._ground_scroll   = 0.0
        self._clouds = [
            {'x': float(random.randint(0, 160)), 'y': float(random.randint(10, 100)),
             'spd': 0.12 + random.random() * 0.2}
            for _ in range(5)
        ]

    def _reset_state(self):
        self.pipes:      list[Pipe]      = []
        self.coins:      list[Coin]      = []
        self.mushrooms:  list[Mushroom]  = []
        self.particles:  list[Particle]  = []
        self.floats:     list[FloatText] = []
        self.score       = 0
        self.coin_count  = 0
        self.combo       = 0
        self.combo_t     = 0.0
        self.triple_t    = 0.0
        self.mush_cd     = 0.0
        self._spawn_x    = pyxel.width + 40.0
        self.elapsed     = 0.0
        self.started     = False
        self.game_over   = False
        # screen-shake
        self._shake      = 0.0
        self._shake_mag  = 0
        # flash
        self._flash_t    = 0.0
        self._flash_col  = WHITE

    def full_reset(self):
        self.bird.reset()
        self._reset_state()

    # ── helpers ───────────────────────────────────────────────────────────

    def _scroll_speed(self):
        t = min(1.0, self.score / 40.0)
        return _lerp(SCROLL_BASE, SCROLL_MAX, t)

    def _pipe_gap(self):
        t = min(1.0, self.score / 35.0)
        return round(_lerp(GAP_START, GAP_MIN, t))

    def shake(self, mag, dur):
        if mag > self._shake_mag:
            self._shake_mag = mag
            self._shake     = dur

    def flash(self, col, dur=0.08):
        self._flash_col = col
        self._flash_t   = dur

    # ── spawn ─────────────────────────────────────────────────────────────

    def _spawn_pipe(self):
        gap   = self._pipe_gap()
        min_t = 22
        max_t = GROUND_Y - gap - 22
        gap_y = random.randint(min_t, max_t)
        pipe  = Pipe(self._spawn_x, gap_y, gap)
        self.pipes.append(pipe)
        self._spawn_coins(pipe)
        if self.mush_cd <= 0 and random.random() < MUSHROOM_CHANCE:
            my = gap_y + gap // 2 + random.randint(-6, 6)
            self.mushrooms.append(Mushroom(self._spawn_x + PIPE_W // 2, my))
            self.mush_cd = MUSHROOM_COOLDOWN
        self._spawn_x += PIPE_SPACING

    def _spawn_coins(self, pipe):
        cx   = pipe.x + PIPE_W // 2
        cy   = pipe.gap_y + pipe.gap_h // 2
        kind = random.choice(['arc', 'line', 'cluster'])
        if kind == 'line':
            n = random.randint(4, 6)
            for i in range(n):
                self.coins.append(Coin(cx - (n-1)*5 + i*10, cy))
        elif kind == 'arc':
            n = 5
            for i in range(n):
                tl = i / (n - 1)
                self.coins.append(Coin(cx - 18 + tl*36,
                                       cy - math.sin(tl * math.pi) * 18))
        else:
            for dx, dy in [(0,0),(-7,-4),(7,-4),(-7,4),(7,4)]:
                self.coins.append(Coin(cx + dx, cy + dy))

    # ── update ────────────────────────────────────────────────────────────

    def update(self, dt):
        self.elapsed += dt
        speed = self._scroll_speed()

        if self.started and not self.game_over:
            if not self.pipes or self.pipes[-1].x < self._spawn_x - PIPE_SPACING:
                self._spawn_pipe()

            for p in self.pipes:    p.update(dt, speed)
            for c in self.coins:    c.update(dt, speed)
            for m in self.mushrooms: m.update(dt, speed)

            self._mountain_scroll += speed * dt
            self._ground_scroll   += speed * dt
            self.mush_cd = max(0.0, self.mush_cd - dt)
            if self.triple_t > 0:
                self.triple_t = max(0.0, self.triple_t - dt)
            if self.combo_t > 0:
                self.combo_t -= dt
                if self.combo_t <= 0:
                    self.combo = 0
            self._spawn_x -= speed * dt

        self.bird.update(dt)

        for p in self.particles:  p.update(dt)
        for f in self.floats:     f.update(dt)
        self.particles = [p for p in self.particles if p.alive]
        self.floats    = [f for f in self.floats    if f.alive]

        # Ground/ceiling
        if self.bird.y + 7 >= GROUND_Y:
            self.bird.y = float(GROUND_Y - 7)
            if self.started and not self.game_over:
                self._kill()
            self.bird.vy = 0
        if self.bird.y < 6:
            self.bird.y = 6.0
            if self.bird.vy < 0:
                self.bird.vy = 0

        if self.started and not self.game_over:
            self._collisions()

        # Clouds
        for cl in self._clouds:
            cl['x'] -= speed * cl['spd'] * dt
            if cl['x'] < -20:
                cl['x'] = pyxel.width + random.randint(0, 30)
                cl['y'] = float(random.randint(10, 100))

        # Prune off-screen
        self.pipes     = [p for p in self.pipes     if p.x + PIPE_W > -4]
        self.coins     = [c for c in self.coins     if c.alive and c.x > -12]
        self.mushrooms = [m for m in self.mushrooms if m.alive and m.x > -12]

        # Shake
        if self._shake > 0:
            self._shake -= dt
            if self._shake <= 0:
                self._shake_mag = 0
        if self._flash_t > 0:
            self._flash_t -= dt

    # ── collisions ────────────────────────────────────────────────────────

    def _collisions(self):
        b = self.bird
        for p in self.pipes:
            if not p.passed and p.x + PIPE_W < b.x:
                p.passed = True
                self.score += 1
            r = p.top_rect
            if b.hits_rect(*r): self._kill(); return
            r = p.bot_rect
            if b.hits_rect(*r): self._kill(); return

        for c in self.coins:
            if c.alive and c.hits(b):
                c.alive = False
                self._collect_coin(c)

        for m in self.mushrooms:
            if m.alive and m.hits(b):
                m.alive = False
                self._collect_mushroom(m)

    def _collect_coin(self, c):
        mult  = 3 if self.triple_t > 0 else 1
        gain  = mult
        self.score      += gain
        self.coin_count += 1
        self.combo      += 1
        self.combo_t     = COMBO_WINDOW
        # Particles
        for _ in range(7):
            ang = random.random() * 2 * math.pi
            sp  = 25 + random.random() * 45
            col = GOLD if random.random() > 0.4 else CREAM
            self.particles.append(
                Particle(c.x, c.y, math.cos(ang)*sp, math.sin(ang)*sp,
                         0.5, col, 1, 60))
        self.floats.append(
            FloatText(c.x, c.y - 6,
                      f'+{gain}', ORNG if mult > 1 else GOLD))
        self.flash(CREAM, 0.045)

    def _collect_mushroom(self, m):
        self.triple_t = TRIPLE_DURATION
        self.shake(2, 0.14)
        self.flash(ORNG, 0.14)
        for i in range(22):
            ang = i / 22 * 2 * math.pi
            sp  = 35 + random.random() * 45
            col = WHITE if i%3==0 else (ORNG if i%2 else CREAM)
            self.particles.append(
                Particle(m.x, m.y, math.cos(ang)*sp, math.sin(ang)*sp,
                         0.75, col, 1, 25))
        self.floats.append(FloatText(m.x, m.y - 10, '3X POWER!', ORNG))

    def _kill(self):
        if self.game_over:
            return
        self.game_over = True
        self.bird.alive = False
        self.shake(6, 0.35)
        self.flash(DKRED, 0.20)
        for _ in range(18):
            ang = random.random() * 2 * math.pi
            sp  = 40 + random.random() * 70
            col = RED if random.random() > 0.5 else ORNG
            self.particles.append(
                Particle(self.bird.x, self.bird.y,
                         math.cos(ang)*sp, math.sin(ang)*sp,
                         1.0, col, 1, 220))

    def world_idle_tick(self, dt):
        """Lightweight tick used on the menu: scrolls clouds + parallax only."""
        speed = SCROLL_BASE * 0.4
        self._mountain_scroll += speed * dt
        self._ground_scroll   += speed * dt
        for cl in self._clouds:
            cl['x'] -= speed * cl['spd'] * dt
            if cl['x'] < -20:
                cl['x'] = pyxel.width + random.randint(0, 30)
                cl['y'] = float(random.randint(10, 100))

    # ── render ────────────────────────────────────────────────────────────

    def draw(self):
        from .sprites import draw_sky, draw_mountains, draw_ground, draw_cloud

        ox = random.randint(-self._shake_mag, self._shake_mag) if self._shake_mag else 0
        oy = random.randint(-self._shake_mag, self._shake_mag) if self._shake_mag else 0

        draw_sky(GROUND_Y)
        for cl in self._clouds:
            draw_cloud(cl['x'] + ox, cl['y'] + oy)
        draw_mountains(self._mountain_scroll, GROUND_Y)
        draw_ground(self._ground_scroll, GROUND_Y)

        for p in self.pipes:     p.draw()
        for c in self.coins:     c.draw()
        for m in self.mushrooms: m.draw()
        for p in self.particles: p.draw()

        self.bird.draw()

        # Mushroom-active aura around bird
        if self.triple_t > 0:
            t = self.elapsed
            r = 12 + int(math.sin(t * 10) * 2)
            pyxel.circb(int(self.bird.x), int(self.bird.y), r,     ORNG)
            pyxel.circb(int(self.bird.x), int(self.bird.y), r + 2, GOLD)

        # Flash overlay
        if self._flash_t > 0 and self._flash_t > 0:
            pyxel.dither(min(0.65, self._flash_t * 8))
            pyxel.rect(0, 0, pyxel.width, pyxel.height, self._flash_col)
            pyxel.dither(1.0)
